GH-4950 LMDB: extensible ID scheme

kenwenzel opened the pull request:
Implement an extensible ID scheme for the LmdbStore that also allows to embed values into IDs.

GitHub issue resolved: #4950 

Briefly describe the changes proposed in this PR:

Use a binary encoding for IDs that allows to represent the value type, if the value itself is embedded within the id and the embedded value itself.

----
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/main/CONTRIBUTING.md) for more details):

 - [x] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [x] I've added tests for the changes I made
 - [x] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/main/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [x] I've [squashed](https://rdf4j.org/documentation/developer/squashing) my commits where necessary 
 - [x] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change

kenwenzel committed:
GH-4950 LMDB: extensible ID scheme

Implement an extensible ID scheme for the LmdbStore that
also allows to embed values into IDs.

hmottestad commented:
The failing integration tests: It's a test that is marked as disabled, but has somehow started to run again. Not sure how that happened.

kenwenzel commented:
@hmottestad It would be good to have this in 5.0.0 as it is a breaking change but I am not sure if I can finish the whole logic.

hmottestad commented:
Sorry. I've already published the M3 milestone build and we now have a feature freeze until the final release. The LMDB Store is experimental, so you should be able to make breaking changes for 5.1.0. Might be better too, since then users can upgrade to 5.0.0 and get all the fixes and enhancements without any major breaking changes.

hmottestad commented:
PS: The develop branch is now 5.1.0 so you can merge into the develop branch without affecting the 5.0.0 release.

nguyenm100 commented:
rdf 1.2 is largely done and jena has an implementation. https://github.com/apache/jena/issues/2805

might be a good idea to brush the dust off of this pr.  how much is left to do @kenwenzel ?

kenwenzel commented:
@nguyenm100 Currently, only the ID scheme in the value store is implemented.
What is missing:
- embedding of (numeric) values into IDs
- handling of triple terms
- indexes for triple terms as part of the triple store

@hmottestad Do we already have an issue for RDF 1.2 and/or can we just adapt the current RDF-star support?

nguyenm100 commented:
Hey @kenwenzel is it possible to just do (1) embedding of values into IDs and merge this pr and then (2) and (3) with rdf 1.2 implementation? or is there some dependency from 1.2 for this to be usable?

kenwenzel commented:
@nguyenm100 Yes, that's perfectly possible.

nguyenm100 commented:
Thanks @kenwenzel are you able to finish (1) and merge this pr? If so, and at that point does it make sense to pull the experimental designation from lmdbstore? (or does that force a major version upgrade?)

kenwenzel committed:
GH-4950 LMDB: support inlined values in value store

kenwenzel committed:
GH-4950 LMDB: support full 64-bit doubles and fix float encoding

kenwenzel commented:
@odysa @nguyenm100 @hmottestad Do we need to make this backwards compatible with current ID scheme or do you think a breaking change is OK?

nguyenm100 commented:
is the perf penalty imposed if it's interacting with lmdb v2 or only present if dealing with lmdb v1?

kenwenzel committed:
GH-4950 LMDB: fix formatting

kenwenzel committed:
GH-4950 LMDB: fix filtering of unused ids

kenwenzel committed:
GH-4950 LMDB: fix handling of doubles and code cleanup

kenwenzel committed:
GH-4950 LMDB: add shared properties file for meta data

kenwenzel committed:
GH-4950 LMDB: fix transaction handling for index creation and reindexing

kenwenzel committed:
GH-4950 LMDB: extensible ID scheme

Implement an extensible ID scheme for the LmdbStore that
also allows to embed values into IDs.

# Conflicts:
#	core/sail/lmdb/src/main/java/org/eclipse/rdf4j/sail/lmdb/ValueStore.java

kenwenzel committed:
GH-4950 LMDB: Do not use Optional<> for member variable

kenwenzel committed:
GH-4950 LMDB: Add config option to disable inlining of literals

kenwenzel commented:
@hmottestad Can I merge this?

kenwenzel committed:
GH-4950 LMDB: Encode doubles with a maximum number of 63 bits to avoid negative ids.

kenwenzel committed:
GH-4950 LMDB: Correctly convert byte values to unsigned integers.

kenwenzel committed:
GH-4950 LMDB: Fix encoding of plain literals without datatypes.
